app/uk/gov/hmrc/tai/model/domain/RateBand.scala

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2024 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.tai.model.domain
+
+import play.api.libs.json.{JsArray, JsValue}
+
+case class RateBand(income: BigDecimal, rate: BigDecimal)
+
+object RateBand {
+  def incomeAndRateBands(json: JsValue): Seq[RateBand] = {
```

Comment:
I think this might be cleaner formulated as a reads. I.e. something like the following (would need to handle all cases):-

  val readsRateBand: Reads[RateBand] =
    (
      (__ \ "income" \ "income").readNullable[BigDecimal] and
        (__ \ "income" \ "rate").readNullable[BigDecimal]
    ).apply { (optIncome, optRate) =>
      (optIncome, optRate) match {
        case (Some(income), Some(rate)) => Some(RateBand(income, rate))
        case _                          => None
      }
    }.filter(_.isDefined)
      .map(_.get)

  val readsSeqRateBand: Reads[Seq[RateBand]] =
    (__ \ "totalLiability" \ "nonSavings" \ "taxBands")
      .readNullable[Seq[RateBand]](Reads.seq(readsRateBand))
      .map(_.getOrElse(Nil))

Reply:
There are a lot of those. The PR is already significant and I am a bit worried of starting to refactor the reads/writes themselves

Reply:
OK, let's leave this one for now